pkgs/build-support/rust/lib/default.nix

```diff
@@ -23,7 +23,9 @@ rec {
 
     ccForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc";
     cxxForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}c++";
-    linkerForHost = if shouldUseLLD stdenv.targetPlatform
+    linkerForHost = if stdenv.targetPlatform.isWasi
+      then "${pkgsBuildHost.llvmPackages.bintools-unwrapped}/bin/${stdenv.cc.targetPrefix}wasm-ld"
```

Comment:
The problem here is that
1. wasm-ld isn't included by bintools-wrapper. And even if add that, rustc seems to pass `-flavor wasm` to the linker, even though its path ends in `wasm-ld`.)
2. `ld.lld` from `bintools-unwrapped` would also work, but from what I understand, it works because rustc passes `-flavor wasm`, which is a legacy option. `ld.lld` from `bintools` chokes on that argument.
3. Using  `"${stdenv.cc}/bin/${stdenv.cc.targetPrefix}clang"` as a linker could work (must be "clang", not "cc" as used in `ccForHost`), but I have been unable to figure out how to explain to clang that the linker it should use is not named `wasm-ld` but `wasm32-unknown-wasi-wasm-ld` (it does have that in its search path already).

Reply:
Wouldn't using a wrapped linker be necessary to find dependencies?  What linker does Rust use by default?

Reply:
> What linker does Rust use by default?

Either `"${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc"` if that's clang or ld.lld otherwise, both are wrapped, from what I can see.
> Wouldn't using a wrapped linker be necessary to find dependencies?

It's possible that there isn't a single rust package that depends on nixpkgs-built dependencies makes it to linking. Most of them do depend on stuff that doesn't work on wasi, like openssl or gui-related libraries. So I can't answer that question. Why using a wrapped linker breaks certainly needs some investigation… Not sure I have time for that.

Reply:
What are the observed symptoms of it not working? Not that if you se NIX_DEBUG than it will also what flags it injects into the build.

Reply:
I would at least leave a comment why we are doing this for now, potentially combined with the error message.

Reply:
@Mic92 Thank you for the `NIX_DEBUG` hint. Getting rust to print the linker command line in the unwrapped case was a bit more tricky, but with both, I was able to confirm what's going on.

The wrapper script was passing hardening flags (`-z relro -z now`) to wasm-ld, which 

* cause lld to not recognize the `-flavor wasm` arguments
* don't make sense on wasm (`warning: unknown -z value: now`) anyway

I'm not sure what's best here, I don't think hardening makes a lot of sense when compiling to wasm anyway, but I don't know of a way to disable hardening for an entire target. What I opted for is to instead add `hardening_unsupported_flags+=" relro bindnow"` in the wrapper script on wasi.

---
For reference:

The output the linker wrapper is:
```
error: linking with `/nix/store/6plp6qym01fbp05lwa1rh4ym50vnw1ng-wasm32-unknown-wasi-llvm-binutils-wrapper-17.0.6/bin/wasm32-unknown-wasi-wasm-ld` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/x86_64-unknown-linux-gnu/bin:/nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/nix/store/708hwm
  = note: HARDENING: disabled flags: pie
          HARDENING: Is active (not completely disabled with "all" flag)
          HARDENING: enabling relro
          HARDENING: enabling bindnow
          extra flags before to /nix/store/5gghgw5dkhw3dx862wynrd26sm3k8xbp-llvm-binutils-17.0.6/bin/wasm32-unknown-wasi-wasm-ld:
            -z
            relro
            -z
            now
          original flags to /nix/store/5gghgw5dkhw3dx862wynrd26sm3k8xbp-llvm-binutils-17.0.6/bin/wasm32-unknown-wasi-wasm-ld:
            -flavor
            wasm
            --rsp-quoting=posix
            --export
            __main_void
            -z
            stack-size=1048576
            --stack-first
            --allow-undefined
            --no-demangle
            /nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/wasm32-wasip1/lib/self-contained/crt1-command.o
            /build/source/target/wasm32-wasip1/release/deps/rq-9dc42ad07b27356d.rq.2f192189ed26cc41-cgu.0.rcgu.o
            -L
            /build/source/target/wasm32-wasip1/release/deps
            -L
            /build/source/target/release/deps
            -L
            /nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/wasm32-wasip1/lib
            -l
            c
            /nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/wasm32-wasip1/lib/libcompiler_builtins-dae8cdcfb0b90f99.rlib
            -L
            /nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/wasm32-wasip1/lib
            -L
            /nix/store/m9dh8g3r6pvbzsxz6b94136l38wmkvgz-wasm32-unknown-wasi-rustc-1.78.0/lib/rustlib/wasm32-wasip1/lib/self-contained
            -o
            /build/source/target/wasm32-wasip1/release/deps/rq-9dc42ad07b27356d.wasm
            --gc-sections
            -O3
          extra flags after to /nix/store/5gghgw5dkhw3dx862wynrd26sm3k8xbp-llvm-binutils-17.0.6/bin/wasm32-unknown-wasi-wasm-ld:
            -L/nix/store/56c3y4h4ba85fxz0876ld9qd8l3ww2hy-libcxx-static-wasm32-unknown-wasi-17.0.6/lib
            -L/nix/store/02cizbm0w4akb46mapwb0in8jcr10srp-wasilibc-static-wasm32-unknown-wasi-21/lib
          wasm32-unknown-wasi-wasm-ld: error: unknown argument: -flavor
          wasm32-unknown-wasi-wasm-ld: warning: unknown -z value: relro
          wasm32-unknown-wasi-wasm-ld: warning: unknown -z value: now
          wasm32-unknown-wasi-wasm-ld: error: cannot open wasm: No such file or directory
```

This is what I was trying with:

```diff
diff --git a/pkgs/build-support/bintools-wrapper/default.nix b/pkgs/build-support/bintools-wrapper/default.nix
index e7fcf173c602..1c955fdd8d0d 100644
--- a/pkgs/build-support/bintools-wrapper/default.nix
+++ b/pkgs/build-support/bintools-wrapper/default.nix
@@ -239,6 +239,8 @@ stdenvNoCC.mkDerivation {
         basename=$(basename "$variant")
         wrap $basename ${./ld-wrapper.sh} $variant
       done
+    '' + optionalString targetPlatform.isWasi ''
+      wrap ${targetPrefix}wasm-ld ${./ld-wrapper.sh} $ldPath/${targetPrefix}wasm-ld
     '';
 
   strictDeps = true;
diff --git a/pkgs/build-support/rust/lib/default.nix b/pkgs/build-support/rust/lib/default.nix
index b52eecccbb2f..7f66bf01c1e5 100644
--- a/pkgs/build-support/rust/lib/default.nix
+++ b/pkgs/build-support/rust/lib/default.nix
@@ -24,7 +24,7 @@ rec {
     ccForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc";
     cxxForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}c++";
     linkerForHost = if stdenv.targetPlatform.isWasi
-      then "${pkgsBuildHost.llvmPackages.bintools-unwrapped}/bin/${stdenv.cc.targetPrefix}wasm-ld"
+      then "${pkgsBuildHost.llvmPackages.bintools}/bin/${stdenv.cc.targetPrefix}wasm-ld"
       else if shouldUseLLD stdenv.targetPlatform
       && !stdenv.cc.bintools.isLLVM
       then "${pkgsBuildHost.llvmPackages.bintools}/bin/${stdenv.cc.targetPrefix}ld.lld"
diff --git a/pkgs/development/tools/rq/default.nix b/pkgs/development/tools/rq/default.nix
index f8ed521ad2cc..589d347dbac5 100644
--- a/pkgs/development/tools/rq/default.nix
+++ b/pkgs/development/tools/rq/default.nix
@@ -24,6 +24,9 @@ rustPlatform.buildRustPackage rec {
     rm build.rs
   '';
 
+  env.NIX_DEBUG = "1";
+  env.RUSTC_LOG = "rustc_codegen_ssa::back::link=info";
+
   VERGEN_SEMVER = version;
 
   meta = with lib; {
```

Reply:
Hardening flags are already platform-dependent, so I think you'd just need to add a case for wasmi.

Reply:
By "by default", I mean what linker does a rustc you'd get from rustup use in this case?

Reply:
Ah, I see. @alyssais For the docker.io/library/rust docker image (which is currently on 1.79) it defaults to `rust-lld -flavor wasm`.

Reply:
(I'm sorry, I force-pushed over this commit again, I mistakenly thought not changing the linker was going to work. But it does have to be set to lld.)